src/frontend/persistent-connection-pair.js

```diff
@@ -0,0 +1,348 @@
+ * @param {number} [options.reconnectDelay=2000] The initial reconnect delay.
+ * @param {number} [options.reconnectDelayCoeff=1.1] The multiplier to apply to the delay for the next reconnect.
+ */
+function PersistentConnectionPair(options) {
+	EventEmitter.call(this);
+	
+	var _this = this;
+	
+	_this._options = _.extend({
+		logPrefix: ''
+	}, options);
+	
+	_this._url = null;
+	
+	_this._connections = [];
+	
+	[ 0, 1 ].forEach(function (index) {
+		var conn = new PersistentConnection(_.extend({}, options, {
+			logPrefix: (options.logPrefix ? options.logPrefix + '[' + (index + 1) + '] ' : undefined)
+		}));
+		
+		var descriptor = {
+			conn: conn,
+			index: index,
+			disabled: false
+		};
+		
+		function addListener(event, handler) {
+			conn.on(event, function () {
+				if (!descriptor.disabled) {
+					var args = [].slice.apply(arguments);
+					handler.apply(_this, [ descriptor ].concat(args));
+				}
+			});
+		}
+		
+		addListener('connecting', _this._onConnConnecting);
+		addListener('connected', _this._onConnConnected);
+		addListener('data', _this._onConnData);
+		addListener('error', _this._onConnError);
+		addListener('end', _this._onConnEnd);
+		
+		_this._connections.push(descriptor);
+	});
+	
+	_this._frontIndex = 1;
+	_this._backIndex = 0;
+	
+	_this.disconnect();
+}
+inherits(PersistentConnectionPair, EventEmitter);
+_.extend(PersistentConnectionPair.prototype, {
+	
+	/**
+	 * @return {boolean}
+	 */
+	isConnecting: function () {
+		return this._isConnecting;
+	},
+	
+	/**
+	 * @return {boolean}
+	 */
+	isConnected: function () {
+		return this._isConnected;
+	},
+	
+	/**
+	 * @return {boolean}
+	 */
+	isReconnecting: function () {
+		return this._isReconnecting;
+	},
+	
+	/**
+	 * Opens a persistent connection on a given URL.
+	 * The server is expected to keep the connection open forever and push the data.
+	 *
+	 * @param {string} url The URL to request. May be updated later via `setUrl`.
+	 */
+	connect: function (url) {
+		var _this = this;
+		
+		_this._url = url;
+		
+		logger.log(_this._options.logPrefix + 'Connect: ' +
+			'back = [' + (_this._backIndex + 1) + '].');
+		
+		_this._connections[_this._backIndex].conn.connect(_this._url);
+	},
+	
+	/**
+	 * Returns the URL.
+	 */
+	getUrl: function () {
+		return this._url;
+	},
+	
+	/**
+	 * Updates the URL for future reconnects.
+	 */
+	setUrl: function (url) {
+		this._url = url;
+	},
+	
+	/**
+	 * Reconnects while having an active connection.
+	 * If there is an active connection, the reconnect happens with a delay.
+	 * If the delay is not needed, use `connect`.
+	 */
+	reconnect: function () {
+		var _this = this;
+		
+		if (!_this._url) {
+			throw new Error('PersistentConnectionPair#reconnect: Missing URL.');
+		}
+		
+		logger.log(_this._options.logPrefix + 'Reconnect: ' +
+			'back = [' + (_this._backIndex + 1) + '].');
+		
+		var conn = _this._connections[_this._backIndex].conn;
+		if (conn.getUrl()) {
```

Comment:
If this `getUrl()` check is only used to verify that the connection has previously been established, I'd prefer a cleaner solution with a dedicated flag / method.